src/pyhf/pdf.py

```diff
@@ -325,40 +332,59 @@ def _create_nominal_and_modifiers(self):
 
         self.mega_mods = mega_mods
 
-        tensorlib, _ = get_backend()
-        thenom = default_backend.astensor(
+        nominal_rates = default_backend.astensor(
             [mega_samples[s]['nom'] for s in self.config.samples]
         )
-        self.thenom = default_backend.reshape(
-            thenom,
+        self._nominal_rates = default_backend.reshape(
+            nominal_rates,
             (
-                1,  # modifier dimension.. thenom is the base
+                1,  # modifier dimension.. nominal_rates is the base
                 len(self.config.samples),
                 1,  # alphaset dimension
                 sum(list(self.config.channel_nbins.values())),
             ),
         )
+        self._nominal_rates = default_backend.tile(
+            self._nominal_rates, (1, 1, self.batch_size or 1, 1)
+        )
 
         self.modifiers_appliers = {
             k: c(
                 [x for x in self.config.modifiers if x[1] == k],  # x[1] is mtype
                 self.config,
                 mega_mods,
+                batch_size=self.batch_size,
                 **self.config.modifier_settings.get(k, {})
             )
             for k, c in modifiers.combined.items()
         }
+        self._precompute()
+        events.subscribe('tensorlib_changed')(self._precompute)
+
+    def _precompute(self):
+        tensorlib, _ = get_backend()
+        self.nominal_rates = tensorlib.astensor(self._nominal_rates)
 
     def expected_auxdata(self, pars):
         tensorlib, _ = get_backend()
         auxdata = None
-        for parname in self.config.auxdata_order:
+        v = ParamViewer(
```

Comment:
Would generally prefer to not have single letter variables.

```suggestion
        view = ParamViewer(
```

Reply:
this needs to be propagated to where we use `v` if we want to change the name

Reply:
Yes, would be a 

```
%s#v\.#view.#g
```

GitHub's suggestions only allow you to make a suggestion on a single line, not across multiple (I think).

Reply:
done